### DISCUSSION TITLE
Custem Fields allow to set json-structured data as default value, but instances using the custom_field can only set strings

### DISCUSSION THREAD

**Discussion by reduzent** [netbox-community/netbox#12065] [Q&A]
I would like to define a  CustomField for the object type Prefix named `dhcp_option_domain_name_servers`  that has the default set to structured data, specifically a list:

```
[
    "10.158.0.53", 
    "10.158.1.53"
]
```

When I access the CustomField `dhcp_option_domain_name_servers` in an export template, I can iterate over the list:

```
name-servers: {{ ', '.join(prefix.cf.get('dhcp_option_domain_name_servers')) }}
```
-->

```
name-servers: 10.158.0.53, 10.158.1.53
```

 However, when I set the CustomField `dhcp_option_domain_name_servers` for a specific instance of a Prefix to the value:

```
[
    "10.157.0.53", 
    "10.157.1.53"
]
```
the value seems to be saved as string and not structured data. The same export template from above returns:

```
name-servers: [, ", 1, 0, ., 1, 5, 7, ., 0, ., 5, 3, ", ,,  , ", 1, 0, ., 1, 5, 8, ., 1, ., 5, 3, ", ]
```

**Comment by reduzent** [comment]
Instances seem to be able to save their custom field data only as string in the database. That's how it looks for an instance that uses the default:

(from  `select custom_field_data from ipam_prefix where id = 1;`)

```
{"dhcp_option_domain_name_servers": ["10.158.0.53", "10.158.1.53"]}
```

That's how it looks for an instance that saved its own value:

(from  `select custom_field_data from ipam_prefix where id = 4;`)

```
{"dhcp_option_domain_name_servers": "[\"10.157.0.53\", \"10.157.1.53\"]"
```

**Comment by reduzent** [comment]
Is this a bug or am I misunderstanding the design of custom fields?

**Reply by kkthxbye-code** [reply]
I think you need to expand more on what you are doing and where it fails.

Creating a JSON custom field and saving, both via. the form and the API, works fine for me. 

> However, when I set the CustomField dhcp_option_domain_name_servers for a specific instance of a Prefix to the value:

What does this mean? What are you doing different here? It's not obvious from your explanation.

**Reply by reduzent** [reply]
Sorry for not being clear. I try again.

This is with 
  * netbox 3.4.6
  * Python 3.9.2

Steps to reproduce the issue 
1. Create new Custom Field:
      - Content Type: `IPAM > Prefix`
      - Name: `dhcp_option_domain_name_servers`
      - Default value: `["10.158.0.53", "10.158.1.53"]`
2. Create Prefix:
     - Prefix: `10.158.0.0/24`
     - Leave custom field `dhcp_option_domain_name_servers` at default value
3. Create another  Prefix:
     - Prefix: `10.157.100.0/24`
     - Set custom field `dhcp_option_domain_name_servers` to `["10.157.0.53", "10.157.1.53"]`
3. Create Export template:
      - Name: `nameservsers`
      - Content Type: `IPAM > Prefix`
      - Template:
```
{%- for prefix in  queryset %}
- prefix: {{ prefix }}
  nameservers: {{ ', '.join(prefix.cf.get('dhcp_option_domain_name_servers')) }}
{% endfor %}
```
4. Access the new export template with curl:
   - `curl -H "Authorization: Token XXX" "https://example.com/api/ipam/prefixes/?export=nameservers"`
 

Result: 

```
- prefix: 10.158.0.0/24
  nameservers: 10.158.0.53, 10.158.1.53

- prefix: 10.157.100.0/24
  nameservers: [, ", 1, 0, ., 1, 5, 7, ., 0, ., 5, 3, ", ,,  , ", 1, 0, ., 1, 5, 7, ., 1, ., 5, 3, ", ]

```

Expected result:

```
- prefix: 10.158.0.0/24
  nameservers: 10.158.0.53, 10.158.1.53

- prefix: 10.157.100.0/24
  nameservers: 10.157.0.53, 10.157.1.53

```

**Reply by kkthxbye-code** [reply]
> Create new Custom Field:
 Content Type: IPAM > Prefix
 Name: dhcp_option_domain_name_servers
 Default value: ["10.158.0.53", "10.158.1.53"]

What type? If you are trying to put json data into a string custom field, that's your issue. If it's a string field, you are also using the default value wrong, it should be `"[\"10.158.0.53\", \"10.158.1.53\"]"` I think. You obviously wouldn't be able to access it as json though, you would have to parse it yourself.

If you are using a JSON custom field (which sounds like is actually what you want) it should work, works fine here at least.

**Reply by reduzent** [reply]
Where can I set the type of the custom field to JSON? 

See here available options:

![image](https://user-images.githubusercontent.com/333373/227974649-dc16dd3d-6699-46c0-84b7-43166a08ba8b.png)

**Reply by kkthxbye-code** [reply]
That field should be scrollable.

![image](https://user-images.githubusercontent.com/400797/227980441-4519e359-584c-4aee-8d51-a0682b8ee6f7.png)

**Reply by reduzent** [reply]
Thanks! That clears everything up for me!

**Comment by reduzent** [comment]
As @kkthxbye-code  suggested correctly, type `JSON` is what I wanted to use. I was simply not seeing that the list of available types is scrollable.  

🤦
